// addons-scanner-utils/dist/functions/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createExpressApp = void 0;
const node_crypto_1 = __importDefault(require("node:crypto"));
const express_1 = __importDefault(require("express"));
const safe_compare_1 = __importDefault(require("safe-compare"));
const error_1 = require("./error");
/**
 * Easily create a new scanner application with built-in authentication
 * middleware. The middleware validates that the content-type is
 * `application/json`, authenticates the request, and ensures the method is
 * POST. Appropriate HTTP error codes are returned for various failure
 * scenarios:
 * - 400 (bad request)
 * - 401 (unauthorized)
 * - 404 (not found)
 * - 405 (method not allowed)
 * - 415 (unsupported media type)
 * - 500 (internal server error)
 *
 * @remarks
 * **Environment Variables:**
 * - `LAMBDA_API_KEY` - Required. API key for authentication (configurable
 *   via `apiKeyEnvVarName`)
 *
 * **Authentication:**
 * - HMAC-SHA256: `Authorization: HMAC-SHA256 <digest>` (digest of request
 *   body)
 */
const createExpressApp = ({ _console = console, _process = process, apiKeyEnvVarName = 'LAMBDA_API_KEY', } = {}) => (handler) => {
    const app = (0, express_1.default)();
    const apiKey = _process.env[apiKeyEnvVarName] || null;
    if (apiKey) {
        // Delete the env var to not expose it to add-ons.
        // eslint-disable-next-line no-param-reassign
        delete _process.env[apiKeyEnvVarName];
    }
    // This is the options we pass to the `json()` middleware.
    const jsonOptions = {
        // This is a hack to retain the raw body on the request object. We need
        // this to verify the signature of the request.
        verify(req, res, buf) {
            req.rawBody = buf;
        },
    };
    // Parse JSON body requests.
    app.use(express_1.default.json(jsonOptions));
    // Authentication middleware plus some extra checks.
    app.use((req, res, next) => __awaiter(void 0, void 0, void 0, function* () {
        var _a;
        // We only allow POST because the `handler` is passed to `app.post()`.
        const allowedMethods = ['POST'];
        if (req.headers['content-type'] !== 'application/json') {
            // We do not throw because we are inside a callback, so we pass an error
            // to the next middleware, which will be the error handler.
            // See: https://expressjs.com/en/guide/error-handling.html
            next((0, error_1.createAppError)({
                message: 'unsupported content type',
                status: 415,
            }));
            return;
        }
        if (!apiKey) {
            next((0, error_1.createAppError)({
                message: `api key must be set`,
                status: 500,
            }));
            return;
        }
        const authorization = (_a = req.get('X-Forwarded-Authorization')) !== null && _a !== void 0 ? _a : req.get('Authorization');
        if (!authorization) {
            next((0, error_1.createAppError)({
                message: `missing authorization header`,
                status: 400,
            }));
            return;
        }
        if (authorization.startsWith('HMAC-SHA256 ') && req.rawBody) {
            const digest = node_crypto_1.default
                .createHmac('sha256', apiKey)
                .update(req.rawBody)
                .digest('hex');
            if (!(0, safe_compare_1.default)(`HMAC-SHA256 ${digest}`, authorization)) {
                next((0, error_1.createAppError)({
                    message: 'authentication has failed',
                    status: 401,
                }));
                return;
            }
        }
        else {
            next((0, error_1.createAppError)({
                message: 'unsupported authorization scheme',
                status: 400,
            }));
            return;
        }
        if (!allowedMethods
            .map((method) => method.toLowerCase())
            .includes(req.method.toLowerCase())) {
            next((0, error_1.createAppError)({ message: 'method not allowed', status: 405 }));
            return;
        }
        next();
    }));
    // We register the handler for the tool that will be exposed.
    app.post('/', handler);
    // NotFound handler.
    app.use((req, res, next) => {
        next((0, error_1.createAppError)({ message: 'not found', status: 404 }));
    });
    // Error handler. Even though we are not using `next`, it must be kept
    // because the Express error handler signature requires 4 arguments.
    app.use(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    (err, req, res, next) => {
        const error = {
            error: err.message,
            extra_info: err.extraInfo || null,
        };
        res.status(err.status || 500).json(error);
        // Also send the error to the cloud provider.
        _console.error(error);
    });
    return app;
};
exports.createExpressApp = createExpressApp;
__exportStar(require("./api"), exports);
__exportStar(require("./auth"), exports);
__exportStar(require("./download"), exports);
__exportStar(require("./error"), exports);
