// snyk/dist/lib/plugins/rubygems/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const inspectors_1 = require("./inspectors");
const missing_targetfile_error_1 = require("../../errors/missing-targetfile-error");
function inspect(root, targetFile) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!targetFile) {
            throw missing_targetfile_error_1.MissingTargetFileError(root);
        }
        const specs = yield gatherSpecs(root, targetFile);
        return {
            plugin: {
                name: 'bundled:rubygems',
                runtime: 'unknown',
            },
            package: {
                name: specs.packageName,
                targetFile: specs.targetFile,
                files: specs.files,
            },
        };
    });
}
exports.inspect = inspect;
function gatherSpecs(root, targetFile) {
    return __awaiter(this, void 0, void 0, function* () {
        for (const inspector of inspectors_1.inspectors) {
            if (inspector.canHandle(targetFile)) {
                return yield inspector.gatherSpecs(root, targetFile);
            }
        }
        throw new Error(`Could not handle file: ${targetFile}`);
    });
}
